\name{variantFilter}
\alias{variantFilter}
%- Also NEED an '\alias' for EACH other topic documented here.
\title{
  %%  ~~function to do ... ~~
  A function to call variants from nextgen sequening data.Works on a tally
  GRanges object as produced by tally2GR in gmapR package
}
\description{
  %%  ~~ A concise (1-5 lines) description of what the function does. ~~
  This function takes a GRanges object as generated by tally2GR in gmapR
  and filters the variants based on a set of options resulting in a set of
  'called' variants.
}
\usage{
variantFilter(granges, useQual = FALSE, pval = 0.05, readCount = 2,
cycleCount = 1, lrt = FALSE, lrt_p = 0.01, lr = TRUE, lower = 0.2,
error = 1/1000, fet_2x2 = TRUE, read_pos = NULL)
}
%- maybe also 'usage' for other objects documented here.
\arguments{
  \item{granges}{
    %%     ~~Describe \code{granges} here~~
    Variants GRanges object as produced by tally2GR in the gmapR package
  }
  \item{useQual}{
    %%     ~~Describe \code{useQual} here~~
    Boolean flag to turn on or off filtering by base quality. If on, the
    LR or LRT will be evaluated on only the high-quality bases
  }
  \item{readCount}{
    minimum number of alternate bases needed to call a variant. The
    default value is 2
  }
  \item{cycleCount}{
    minimum number of cycles that the variant base must be seen in to
    call. The default value is 1.
  }
  \item{pval}{
    %%     ~~Describe \code{pval} here~~
    pval to be used in filtering during the Fisher's exact test
  }
  \item{lrt}{
    %%     ~~Describe \code{lrt} here~~
    Boolean flag to choose whether to use the LRT. If this is set to
    true, lr must be set to false.
  }
  \item{lrt_p}{
    %%     ~~Describe \code{lrt.p} here~~
    p-value threshold to use when the LRT method is used.
  }
  \item{lr}{
    %%     ~~Describe \code{lr} here~~
    Boolean flag to turn on the LR method. 
  }
  \item{lower}{
    %%     ~~Describe \code{lower} here~~
    Used in the LR method this is the lowest frequency you with to have
    power to detect.
  }
  \item{error}{
    %%     ~~Describe \code{error} here~~
    assumed error rate from the sequencing data (eg 1/1000)
  }
  \item{read_pos}{
    %%     ~~Describe \code{read_pos} here~~
    gives the name of the bin to filter on for the read position filter
  }
  \item{fet_2x2}{
    %%     ~~Describe \code{fet_2x2} here~~
    Boolean flag to turn on or off the Fisher's exact test of strand
    usage bias in the variants
  }
}
\value{
%%  ~Describe the value returned
%%  If it is a LIST, use
%%  \item{comp1 }{Description of 'comp1'}
%%  \item{comp2 }{Description of 'comp2'}
%% ...
\item{raw_granges }{
  GRanges object of the raw variants. This is all
  positions with 1 or more non-reference bases in the bam file.
}
\item{filtered_granges }{
  GRanges object of the filtered (called) variants.
}
\item{N_nucleotide_rejects }{
  number of positions removed due to the non-ref base being an N
}
\item{low_count_rejects }{
  number of positions removed due to having too few non-reference
  bases (too few is defined as less than 2 by default)
}
\item{frequency_rejects }{
  number of positions removed due to either failing the LRT or
  having frequency below the threshold set in the LR
}
\item{fisher_test_rejects_2by2 }{
  number of positions removed for failing the Fisher's exact
  test on the strand usage of the variant.
}
\item{read_pos_rejects }{
  number of positions removed for failing the read positional filter
  this will be 0 if read_pos is NULL
}
}
\author{
  %%  ~~who you are~~
  Jeremiah Degenhardt
}
